dali/operators/decoder/audio/audio_decoder_op.h

```diff
@@ -0,0 +1,135 @@
+
+
+
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_DECODER_AUDIO_AUDIO_DECODER_OP_H_
+#define DALI_OPERATORS_DECODER_AUDIO_AUDIO_DECODER_OP_H_
+
+#include <memory>
+#include <string>
+#include <vector>
+#include "dali/core/static_switch.h"
+#include "dali/operators/decoder/audio/audio_decoder.h"
+#include "dali/operators/decoder/audio/generic_decoder.h"
+#include "dali/pipeline/data/backend.h"
+#include "dali/pipeline/workspace/workspace.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/pipeline/workspace/host_workspace.h"
+
+namespace dali {
+
+namespace detail {
+
+const std::string kOutputTypeName = "dtype";  // NOLINT
+const int kNumOutputs = 2;
+
+}  // namespace detail
+
+class AudioDecoderCpu : public Operator<CPUBackend> {
+ private:
+  using Backend = CPUBackend;
+
+ public:
+  explicit inline AudioDecoderCpu(const OpSpec &spec) :
+          Operator<Backend>(spec),
+          output_type_(spec.GetArgument<DALIDataType>(detail::kOutputTypeName)) {}
+
+
+  inline ~AudioDecoderCpu() override = default;
+
+ protected:
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<Backend> &ws) override {
+    auto &input = ws.template InputRef<Backend>(0);
+    const auto batch_size = input.shape().num_samples();
+
+    for (int i = 0; i < input.shape().num_samples(); i++) {
+      DALI_ENFORCE(input.shape()[i].size() == 1, "Input must be 1D encoded byte data");
+    }
+    DALI_ENFORCE(IsType<uint8_t>(input.type()), "Input must be stored as uint8_t data.");
+    TypeInfo type;
+    TypeInfo type_i32;
+    type_i32.SetType<int>(DALI_INT32);
+    decoders_.resize(batch_size);
+    sample_meta_.resize(batch_size);
+    files_names_.resize(batch_size);
+
+    TYPE_SWITCH(output_type_, type2id, OutputType, (int16_t, int32_t, float), (
+            for (int i=0; i < batch_size; i++) {
```

Comment:
bad indentation

Reply:
I did it on purpose, to make this TYPE_SWITCH thing more readable: that way you see clearer, what is actually the type-switched code ;)